### PR TITLE
Day 06 - Add Basic Search and Pagination to GET Transactions

### PR DESCRIPTION
**Pull request by kishanpatel789**
## Change Log
- Enhance GET transaction route with search feature
  - Sort transactions returned
  - Add pagination using "page number" and "page size" pattern
  - Add ability to filter transactions by fields Vendor, Category, and Note
- Implement changes recommended in [PR#18](https://github.com/kishanpatel789/finance_tracker/pull/18) (branch `day05`)
  - Update dummy data script with environment variables and more accurate function names
  - Remove metadata from variable names of re-used Field objects
  - Further parametrize validation tests

## TODO
- [ ] Further enhance GET transaction route
  - [ ] Implement filter by date range (on transaction date)
  - [ ] Enhance pagination for edge cases (eg input page number is too high, etc)
- [ ]  Set up Postgres database in Docker to replace current SQLite database

### CODE REVIEW

File: scripts/generate_dummy_data.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 from datetime import date, datetime, timedelta, timezone
 from decimal import Decimal
 
+from decouple import config
```
Comment (bbelderbos) [inline scripts/generate_dummy_data.py:5]: great choice

File: scripts/generate_dummy_data.py [hunk 2]
```diff
@@ -53,11 +54,12 @@ def generate_transactions(categories: list[Category], n=250) -> list[Transaction
 
 
 def get_engine():
-    engine = create_engine("sqlite:///finance_tracker.sqlite", echo=False)
+    DATABASE_URL = config("DATABASE_URL_DUMMY", "sqlite:///finance_tracker.sqlite")
+    engine = create_engine(DATABASE_URL, echo=False)
```
Comment (bbelderbos) [inline scripts/generate_dummy_data.py:58]: I think echo=False is the default, if you have a DEBUG bool you could also couple it to that: `echo=DEBUG`

File: tests/test_api_categories.py [hunk 3]
```diff
@@ -56,51 +56,41 @@ def test_create_category_422_empty_payload(client: TestClient):
...
     data = response.json()
     assert response.status_code == 422
-    assert data["detail"][0]["type"] == "string_too_long"
+    assert data["detail"][0]["type"] == expected_error_type
 
 
 @pytest.mark.parametrize(
     "method,url", [("POST", "/categories"), ("PATCH", "/categories/1")]
 )
+@pytest.mark.parametrize(
```
Comment (bbelderbos) [inline tests/test_api_categories.py:79]: nice, did clean it up quite a bit

File: tests/test_api_transactions.py [hunk 4]
```diff
@@ -91,81 +91,87 @@ def test_create_transaction_422(client: TestClient):
 @pytest.mark.parametrize(
     "method,url", [("POST", "/transactions"), ("PATCH", "/transactions/1")]
 )
+@pytest.mark.parametrize(
+    "payload,expected_error_type",
```
Comment (bbelderbos) [inline tests/test_api_transactions.py:95]: ruff (black) auto-format is so nice here, can you believe there was a time I would do this manually?! 😱

File: tests/test_api_transactions.py [hunk 5]
```diff
@@ -174,6 +180,21 @@ def test_get_transactions(client: TestClient, add_transaction, add_another_trans
 
     assert response.status_code == 200
     assert len(data) == 2
+    assert data[0]["trans_date"] == "2025-11-02"
+    assert data[1]["trans_date"] == "2024-07-14"
```
Comment (bbelderbos) [inline tests/test_api_transactions.py:184]: still no need for freezing datetimes because we hardcode them right?

File: tests/test_api_transactions.py [hunk 6]
```diff
@@ -174,6 +180,21 @@ def test_get_transactions(client: TestClient, add_transaction, add_another_trans
...
+
+
+@pytest.mark.parametrize(
+    "term,expected_count",
+    [("Utilities", 1), ("%20%20KrOgER%20", 1), ("DoesNotExist", 0)],
+)
+def test_get_transactions_search_by_term(
+    client: TestClient, add_transaction, add_another_transaction, term, expected_count
```
Comment (bbelderbos) [inline tests/test_api_transactions.py:192]: love how fixtures + parametrize make test code so much shorter

File: src/models.py [hunk 7]
```diff
@@ -12,15 +12,15 @@ class MySQLModel(SQLModel):
...
 
 
-field_money_dec_or_none = Field(max_digits=10, decimal_places=2, default=None)
-field_money_dec = Field(max_digits=10, decimal_places=2)
-field_str_1_25_or_none = Field(min_length=1, max_length=25, default=None)
-field_str_1_25 = Field(min_length=1, max_length=25)
+MoneyFieldOrNone = Field(max_digits=10, decimal_places=2, default=None)
+MoneyField = Field(max_digits=10, decimal_places=2)
+StringFieldOrNone = Field(min_length=1, max_length=25, default=None)
+StringField = Field(min_length=1, max_length=25)
```
Comment (bbelderbos) [inline src/models.py:18]: much better

File: src/routers/transactions.py [hunk 8]
```diff
@@ -37,8 +39,42 @@ def create_transaction(transaction: TransactionCreate, session: SessionDep):
 
 
 @router.get("/", response_model=list[TransactionRead])
-def read_transactions(session: SessionDep):
-    transactions = session.exec(select(Transaction)).all()
+def read_transactions(
+    session: SessionDep,
+    pagination_input: PaginationDep,
```
Comment (bbelderbos) [inline src/routers/transactions.py:44]: I wonder if this is an abstraction too much? Now I have to go look what `PaginationDep`has.

Compare that to docs example:
https://sqlmodel.tiangolo.com/tutorial/fastapi/limit-and-offset/#add-a-limit-and-offset-to-the-query-parameters

```
def read_heroes(offset: int = 0, limit: int = Query(default=100, le=100)):
```

This is more explicit and you can still add validation there via `Query`

What do you think?

File: src/routers/transactions.py [hunk 9]
```diff
@@ -37,8 +39,42 @@ def create_transaction(transaction: TransactionCreate, session: SessionDep):
...
+    q: Annotated[str | None, Query(max_length=40)] = None,
+):
+    query = select(Transaction)
+
+    # search filter
+    if q is not None:
+        q = q.strip().lower()
+        q = re.sub(r"\s+", " ", q)
```
Comment (bbelderbos) [inline src/routers/transactions.py:52]: could do in one go: `q = re.sub(r"\s+", " ", q.lower())`

File: src/routers/transactions.py [hunk 10]
```diff
@@ -37,8 +39,42 @@ def create_transaction(transaction: TransactionCreate, session: SessionDep):
...
 
 @router.get("/", response_model=list[TransactionRead])
-def read_transactions(session: SessionDep):
-    transactions = session.exec(select(Transaction)).all()
+def read_transactions(
+    session: SessionDep,
+    pagination_input: PaginationDep,
+    q: Annotated[str | None, Query(max_length=40)] = None,
+):
+    query = select(Transaction)
```
Comment (bbelderbos) [inline src/routers/transactions.py:47]: checking for n+1 issues, all good, I think the fact you have `category: Category | None = Relationship(..., sa_relationship_kwargs={"lazy": "selectin"})` we won't have redundant/extra queries here 📈

**Review by bbelderbos — APPROVED** [review]
Looks great! What's next? Any blockers?

File: src/models.py (hunk 7 above)
Comment (kishanpatel789) [reply src/models.py:18]: Thanks for the tip!

File: src/routers/transactions.py (hunk 9 above)
Comment (kishanpatel789) [reply src/routers/transactions.py:52]: Yeah, I went back on forth on this one. I think you forgot `strip()` in your snippet. 

The full one-liner would look like this: `q = re.sub(r"\s+", " ", q.strip().lower())`
It's not terrible, but it's getting a bit too long for readability. I also liked the separation between built-in `str` cleanup (strip and lower) and secondary `re` cleanup.

File: tests/test_api_categories.py (hunk 3 above)
Comment (kishanpatel789) [reply tests/test_api_categories.py:79]: Absolutely!

File: src/routers/transactions.py (hunk 8 above)
Comment (kishanpatel789) [reply src/routers/transactions.py:44]: Good thoughts! `read_transactions` is a work in progress and I'm still thinking about what the final form will be. A lot of this gets into API design. 

I started the pagination implementation using the `offset`/`limit` example in the FastAPI docs. But after using it a bit, I was dissatisfied. I like the `page`/`page_size` approach because it makes a bit more intuitive sense for the API user. The user always knows what "page" they're on and can increment by +1 (`new_page = old_page + 1`). The  `offset`/`limit` is a bit more complicated (`new_offset` = `old_offset` + `old_limit`). I'm splitting hairs here.

Anyway, I'm thinking about expanding the response model `read_transactions` to include metadata, like the total record account available in the search query, the API URLs (or links) to navigate to the next or previous page, etc.  

Some of these ideas are from consuming different APIs as a data engineer. I enjoyed well-designed responses that helped make multiple API calls. Here's an API I played with a few months ago that has the `page`/`page_size` and "links" design: https://docs.potterdb.com/apis/rest#pagination

File: src/routers/transactions.py (hunk 10 above)
Comment (bbelderbos) [reply src/routers/transactions.py:47]: not sqlmodel but still relevant for the concept:

[video] Optimizing Django Queries: How to Solve the N+1 Problem with select_related()
https://www.youtube.com/watch?v=e_8JvcP1q48

File: src/routers/transactions.py (hunk 8 above)
Comment (bbelderbos) [reply src/routers/transactions.py:44]: oh wow nice example, didn't realize you could use it like that then in the url params, nice
leave this then like this for now

Comment (bbelderbos) [reply src/routers/transactions.py:44]: and awesome the care you put in the design 📈
